core/src/main/java/jenkins/install/SetupWizard.java

```diff
@@ -0,0 +1,180 @@
+                bc.commit();
+            }
+        }
+        else {
+            admin = j.getUser(SetupWizard.initialSetupAdminUserName);
+        }
+        
+        String setupKey = null;
+        if(admin != null && admin.getProperty(SetupWizard.AuthenticationKey.class) != null) {
+            setupKey = admin.getProperty(SetupWizard.AuthenticationKey.class).getKey();
+        }
+        if(setupKey != null) {
+            LOGGER.info("\n\n*************************************************************\n"
+                    + "*************************************************************\n"
+                    + "*************************************************************\n"
+                    + "\n"
+                    + "Jenkins initial setup is required. A security token is required to proceed. \n"
+                    + "Please use the following security token to proceed to installation: \n"
+                    + "\n"
+                    + "" + setupKey + "\n"
+                    + "\n"
+                    + "*************************************************************\n"
+                    + "*************************************************************\n"
+                    + "*************************************************************\n");
+        }
+        
+        try {
+            PluginServletFilter.addFilter(FORCE_SETUP_WIZARD_FILTER);
+        } catch (ServletException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    /**
+     * Remove the setupWizard filter, ensure all updates are written to disk, etc
+     */
+    public void doCompleteSetupWizard() {
+        InstallUtil.saveLastExecVersion();
+        try {
+            PluginServletFilter.removeFilter(FORCE_SETUP_WIZARD_FILTER);
+        } catch (ServletException e) {
+            throw new AssertionError(e); // never happen because our Filter.init is no-op
+        }
+    }
+
+    // Stores a user property for the authentication key, which is really the auto-generated user's password
+    public static class AuthenticationKey extends UserProperty {
+        String key;
+        
+        public AuthenticationKey() {
+        }
+        
+        public AuthenticationKey(String key) {
+            this.key = key;
+        }
+        
+        public String getKey() {
+            return key;
+        }
+        
+        public void setKey(String key) {
+            this.key = key;
+        }
+    }
+    
+    /**
+     * This filter will validate that the security token is provided
+     */
+    private final Filter FORCE_SETUP_WIZARD_FILTER = new Filter() {
+        @Override
+        public void init(FilterConfig cfg) throws ServletException {
+        }
+
+        @Override
+        public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+            // As an extra measure of security, the install wizard generates a security token, and
```

Comment:
What is the extra security, exactly? You have already configured a restrictive security realm and authorization strategy. Before that takes effect, you get a 503 served from the _Please wait while Jenkins is getting ready to work_ page. After it takes effect,

    curl -i -u initial-setup-admin-user:…token… http://localhost:8080/api/json

works, as does logging in with the security token and then browsing to *any other* page URL in Jenkins besides the root—including, say, `/configureSecurity/`.

So it seems like this filter adds nothing to security. It merely guides the user to the setup wizard. Which could be accomplished more simply by just having `Jenkins/login.jelly` redirect you to `/setupWizard/`.